SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipLiveness.swift

```diff
@@ -0,0 +1,977 @@
+         is InitExistentialAddrInst, is OpenExistentialAddrInst,
+         is IndexAddrInst, is MarkUnresolvedNonCopyableValueInst,
+         is ProjectBlockStorageInst, is TailAddrInst, is StoreBorrowInst,
+         is UncheckedAddrCastInst, is MarkUninitializedInst, is DropDeinitInst,
+         is TuplePackElementAddrInst, is CopyableToMoveOnlyWrapperAddrInst,
+         is MoveOnlyWrapperToCopyableAddrInst:
+      let svi = address.instruction as! SingleValueInstruction
+      return walkDown(projection: svi, of: address)
+
+    case is ReturnInst, is ThrowInst, is YieldInst, is TryApplyInst,
+         is SwitchEnumAddrInst, is CheckedCastAddrBranchInst,
+         is SelectEnumAddrInst, is InjectEnumAddrInst,
+         is LoadInst, is CopyAddrInst, is StoreInst, is DestroyAddrInst,
+         is LoadUnownedInst, is StoreUnownedInst,
+         is LoadWeakInst, is StoreWeakInst,
+         is AssignInst, is AssignByWrapperInst, is AssignOrInitInst,
+         is MarkUnresolvedMoveAddrInst, is ExplicitCopyAddrInst,
+         is RetainValueAddrInst, is ReleaseValueAddrInst,
+         is UncheckedRefCastAddrInst, is UnconditionalCheckedCastAddrInst,
+         is TupleAddrConstructorInst, is InitBlockStorageHeaderInst,
+         is DeallocStackInst, is EndApplyInst,
+         is IsUniqueInst, is MarkFunctionEscapeInst,
+         is WitnessMethodInst, is ValueMetatypeInst, is ExistentialMetatypeInst,
+         is DeinitExistentialAddrInst,
+         is GetAsyncContinuationAddrInst, is KeyPathInst,
+         is PackElementSetInst, is PackElementGetInst:
+      return leafUse(address: address)
+
+    case let builtin as BuiltinInst:
+      switch builtin.id {
+      case .TSanInoutAccess, .ResumeThrowingContinuationReturning,
+           .ResumeNonThrowingContinuationReturning, .Copy, .GenericAdd,
+           .GenericFAdd, .GenericAnd, .GenericAShr, .GenericLShr, .GenericOr,
+           .GenericFDiv, .GenericMul, .GenericFMul, .GenericSDiv,
+           .GenericExactSDiv, .GenericShl, .GenericSRem, .GenericSub,
+           .GenericFSub, .GenericUDiv, .GenericExactUDiv, .GenericURem,
+           .GenericFRem, .GenericXor, .TaskRunInline, .ZeroInitializer,
+           .GetEnumTag, .InjectEnumTag:
+        return leafUse(address: address)
+      default:
+        // TODO: SIL verification should check that this exhaustively
+        // recognizes all builtin address uses.
+        return .abortWalk
+      }
+
+    case is AddressToPointerInst:
+      return pointerEscape(address: address)
+
+    case let pai as PartialApplyInst where !pai.isOnStack:
+      return pointerEscape(address: address)
+
+    case is BranchInst, is CondBranchInst:
+      fatalError("address phi is not allowed")
+
+    default:
+      if address.instruction.isIncidentalUse {
+        return leafUse(address: address)
+      }
+      // Unkown instruction.
+      return unknown(address: address)
+    }
+  }
+}
+
+/// Enumerate all special cases of ownership uses in a visitor
```

Comment:
It's not clear to me what `OwnershipUseVisitor` does or should do. Also this comment is not really useful.
On the one hand it appears to wrap a visitor around the `OperandOwnership` enum. On the other hand it looks like it's also doing some kind of def-use walk.
The confusion already starts with that I'm not sure what the terms "outer" and "inner" scope/liferange/value actually refer to.

Reply:
`OwnershipUseVisitor` is the one place that define the effects of operand ownership. We need this so we don't end up with ad-hoc utilities in OptUtils like `InstructionRange.insert(borrowScopeOf)` which is out of place, incomplete, and likely misused.

Reply:
Still, I don't know what an `OwnershipUseVisitor` does and what I can use it for.
Small and simple utilities like `InstructionRange.insert(borrowScopeOf)` are much easier to use because it's immediately clear what they do.
We really need to keep those new utilities understandable and easy to use. Part of it is good API names and good comments. If the purpose of a utility cannot be expressed in a few sentences in a comment its design is probably too complex.